### PR TITLE
fix(resolve): prevent infinite loop when glob-import self

### LINKED ISSUE #110164: ICE: stack overflow in resolver
### Code

```Rust
use crate::*;

fn main() {
    use _::*;
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (696aaad58 2023-04-09)
binary: rustc
commit-hash: 696aaad58c57a589f6fb2ecff5bae2eec581cb71
commit-date: 2023-04-09
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 16.0.2
```

### Error output

```
<output>
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: expected identifier, found reserved identifier `_`
 --> /tmp/treereduce.out:4:9
  |
4 |     use _::*;
  |         ^ expected identifier, found reserved identifier


Thread 2 "rustc" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe9dff6c0 (LWP 3656741)]
0x00007ffff534e571 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
(gdb) bt
#0  0x00007ffff534e571 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#1  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#2  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#3  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#4  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#5  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#6  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#7  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#8  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#9  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#10 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#11 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#12 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#13 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#14 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#15 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#16 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#17 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#18 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#19 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#20 0


.....
```

</p>
</details>

### PR DESCRIPTION
**Pull request by bvanjoi**
close #110164

### CODE REVIEW

File: tests/ui/underscore-imports/issue-110164.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+//~| ERROR unresolved import `_`
+use _::*;
+//~^ ERROR expected identifier, found reserved identifier `_`
+//~| ERROR unresolved import `_`
+
+fn main() {
+    use _::a;
+    //~^ ERROR expected identifier, found reserved identifier `_`
+    //~| ERROR unresolved import `_`
+    use _::*;
```
Comment (petrochenkov) [inline tests/ui/underscore-imports/issue-110164.rs:16]: Why does `use _::*` trigger the "glob-import a module into itself" in the first place?
I'd expect `_` to resolve into `Res::Err` here.

Is it possible to reproduce the ICE without `_::*`, just with regular well-formed imports?

File: compiler/rustc_resolve/src/ident.rs [hunk 2]
```diff
@@ -1027,15 +1027,21 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
                 continue;
             }
-            let module = match glob_import.imported_module.get() {
+            let imported_module = match glob_import.imported_module.get() {
                 Some(ModuleOrUniformRoot::Module(module)) => module,
                 Some(_) => continue,
                 None => return Err((Undetermined, Weak::Yes)),
             };
+
+            if ptr::eq(imported_module, module) {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:1036]: The "cannot glob-import a module into itself" error is reported for the case `glob_import.imported_module == glob_import.parent_scope.module`.
I'm not sure `module` here is always the same as `glob_import.parent_scope.module`, it needs at least an assert.

Or just check this condition on `glob_import.parent_scope.module` directly.

File: tests/ui/underscore-imports/issue-110164.rs (hunk 1 above)
Comment (bvanjoi) [reply tests/ui/underscore-imports/issue-110164.rs:16]: > Is it possible to reproduce the ICE without _::*, just with regular well-formed imports?

Will not occur, other cases will hit [map_err](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/ident.rs#L870)

> I'd expect _ to resolve into Res::Err here.

When executing [new_key](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/ident.rs#L870), [special processing](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/lib.rs#L1594) is performed for `_` to ensure that `_` each time the `key` is is not the same, it is intended to solve the case of [`xxx as _`](https://github.com/rust-lang/rust/blob/master/tests/ui/underscore-imports/basic.rs#L33)

Comment (petrochenkov) [reply tests/ui/underscore-imports/issue-110164.rs:16]: Haha, fascinating.
The `underscore_disambiguator += 1` should only apply when we define names, not when we are retrieving already added definitions.

Comment (petrochenkov) [reply tests/ui/underscore-imports/issue-110164.rs:16]: I suggest adding `BindingKey::new` method that will be used everywhere by default, and rename `new_key` to `new_disambiguated_key` and use it only when adding `use foo as _` imports (in `fn add_import` I guess, but I'm not sure).

Note that if `new_key` is called for a single import twice in different contexts, then the resulting disambiguators are not consistent with each other, and it's something that is suspicious at least.

Comment (petrochenkov) [reply tests/ui/underscore-imports/issue-110164.rs:16]: This change is independent from glob imports and we should do it first.

The glob change also looks reasonable, maybe it fixes some other actual issues.

Comment (bvanjoi) [reply tests/ui/underscore-imports/issue-110164.rs:16]: Then I will open a new PR to rewrite `new_key`, what do you think?

Comment (petrochenkov) [reply tests/ui/underscore-imports/issue-110164.rs:16]: Yes, new PR.

**Comment by JohnCSimon** [comment]
@bvanjoi 
ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by bvanjoi** [comment]
this issue had been solved by another PR, so it can be closed.

**Pull request closed without merging**